R/prof.to.features.R

```diff
@@ -1,622 +1,706 @@
+      m <- sum(x * y) / sum(y)
+      s1 <- s2 <- sum(y * (x - m)^2) / sum(y)
+      scale <- sum(y) / s1
+      to.return <- c(m, s1, s2, scale)
     }
-    return(to.return)
+  }
+  return(to.return)
 }
 
+#' @importFrom magittr %>%
+#' @importFrom dplyr filter arrange
 #' @export
-bigauss.mix <- function(x, y, power = 1, do.plot = FALSE, sigma.ratio.lim = c(0.1, 10), bw = c(15, 30, 60), eliminate = .05, max.iter = 25, estim.method, BIC.factor = 2) {
-    all.bw <- bw[order(bw)]
+bigauss.mix <- function(chr_profile, power = 1, do.plot = FALSE, sigma.ratio.lim = c(0.1, 10), bw = c(15, 30, 60), eliminate = .05, max.iter = 25, estim.method, BIC.factor = 2) {
+
+  all.bw <- bw[order(bw)]
+  results <- new("list")
+  smoother.pk.rec <- smoother.vly.rec <- new("list")
+  bic.rec <- all.bw
+
+  if (do.plot) {
+    par(mfrow = c(ceiling(length(all.bw) / 2), 2))
+    par(mar = c(1, 1, 1, 1))
+  }
+
+  last.num.pks <- Inf
+
+  chr_profile_unfiltered <- chr_profile
+  chr_profile <- data.frame(chr_profile) %>%
+    filter(intensity > 1e-5) %>%
+    arrange(base_curve)
+
+  for (bw.n in length(all.bw):1)
+  {
+    bw <- all.bw[bw.n]
+    this.smooth <- ksmooth(chr_profile_unfiltered[, "base_curve"], chr_profile_unfiltered[, "intensity"], kernel = "normal", bandwidth = bw)
+    turns <- find.turn.point(this.smooth$y)
+    pks <- this.smooth$x[turns$pks]
+    vlys <- c(-Inf, this.smooth$x[turns$vlys], Inf)
+
+    smoother.pk.rec[[bw.n]] <- pks
+    smoother.vly.rec[[bw.n]] <- vlys
+    if (length(pks) != last.num.pks) {
+      last.num.pks <- length(pks)
+      l <- length(chr_profile[, "base_curve"])
+      dx <- c(chr_profile[2, "base_curve"] - chr_profile[1, "base_curve"], (chr_profile[3:l, "base_curve"] - chr_profile[1:(l - 2), "base_curve"]) / 2, chr_profile[l, "base_curve"] - chr_profile[l - 1, "base_curve"])
+      if (l == 2) {
+        dx <- rep(diff(chr_profile[, "base_curve"]), 2)
+      }
+
+      # initiation
+      m <- s1 <- s2 <- delta <- pks
+      for (i in 1:length(m))
+      {
+        sel.1 <- which(chr_profile[, "base_curve"] >= max(vlys[vlys < m[i]]) & chr_profile[, "base_curve"] < m[i])
+        s1[i] <- sqrt(sum((chr_profile[sel.1, "base_curve"] - m[i])^2 * chr_profile[sel.1, "intensity"] * dx[sel.1]) / sum(chr_profile[sel.1, "intensity"] * dx[sel.1]))
+
+        sel.2 <- which(chr_profile[, "base_curve"] >= m[i] & chr_profile[, "base_curve"] < min(vlys[vlys > m[i]]))
+        s2[i] <- sqrt(sum((chr_profile[sel.2, "base_curve"] - m[i])^2 * chr_profile[sel.2, "intensity"] * dx[sel.2]) / sum(chr_profile[sel.2, "intensity"] * dx[sel.2]))
+
+        delta[i] <- (sum(chr_profile[sel.1, "intensity"] * dx[sel.1]) + sum(chr_profile[sel.2, "intensity"] * dx[sel.2])) / ((sum(dnorm(chr_profile[sel.1, "base_curve"], mean = m[i], sd = s1[i])) * s1[i] / 2) + (sum(dnorm(chr_profile[sel.2, "base_curve"], mean = m[i], sd = s2[i])) * s2[i] / 2))
+      }
+      delta[is.na(delta)] <- 1e-10
+      s1[is.na(s1)] <- 1e-10
+      s2[is.na(s2)] <- 1e-10
```

Comment:
Seems like a good candidate for a dataframe again - also see the exclusion section below.